src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonAfterTypeMemberDeclarationCheck.java

```diff
@@ -0,0 +1,158 @@
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.Arrays;
+import java.util.List;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
+
+/**
+ * <p>
+ * Checks if unnecessary semicolon is used after code type member declaration.
+ * The check will flag the following with warnings:
+ * </p>
+ * <pre>
+ * class A {
+ *     ; // standalone semicolon
+ *     {}; // semicolon after init block
+ *     static {}; // semicolon after static init block
+ *     A(){}; // semicolon after constructor definition
+ *     void method() {}; // semicolon after method definition
+ *     int field = 10;; // semicolon after field declaration
+ * }
+ * </pre>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;UnnecessarySemicolonAfterTypeMemberDeclaration&quot;/&gt;
+ * </pre>
+ *
+ * @since 8.23
+ */
+@StatelessCheck
+public final class UnnecessarySemicolonAfterTypeMemberDeclarationCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_SEMI = "unnecessary.semicolon";
+
+    /**
+     * List of tokens that is used to check if token is a class member.
+     *
+     * @see #isClassMember(DetailAST)
+     */
+    private static final List<Integer> CLASS_MEMBER_TOKENS = Arrays.asList(
+        TokenTypes.STATIC_INIT,
+        TokenTypes.INSTANCE_INIT,
+        TokenTypes.CTOR_DEF,
+        TokenTypes.METHOD_DEF
+    );
+
+    /**
+     * List of tokens that is used to check if token is a nested type.
+     *
+     * @see #isNestedTypeDeclaration(DetailAST)
+     */
+    private static final List<Integer> TYPE_TOKENS = Arrays.asList(
+        TokenTypes.CLASS_DEF,
+        TokenTypes.INTERFACE_DEF,
+        TokenTypes.ENUM_DEF,
+        TokenTypes.ANNOTATION_DEF
+    );
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.SEMI,
```

Comment:
I am a bit confused that this check will be called on almost every line of code.

Reply:
How it can be changed?

Reply:
I will try to change it to inspect only necessary tokens, not every semicolon. I agree, it looks like overkill.

Reply:
@pbludov I changed implementation, please take a look.